### Problem Statement :- Predict the Sales prices 

In [12]:
import pandas as pd 
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score
import matplotlib.pyplot 
import seaborn as sns 

--------------------------------------------------------------------

### Data science Stage 2 :- Data Collection ( stage 1 is understanding the problem / need / requirement ) 

In [2]:
train_data = pd.read_csv("train.csv")

train_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave   NaN      Reg   
1        2          20       RL         80.0     9600   Pave   NaN      Reg   
2        3          60       RL         68.0    11250   Pave   NaN      IR1   
3        4          70       RL         60.0     9550   Pave   NaN      IR1   
4        5          60       RL         84.0    14260   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave   NaN      Reg   
1456  1457          20       RL         85.0    13175   Pave   NaN      Reg   
1457  1458          70       RL         66.0     9042   Pave   NaN      Reg   
1458  1459          20       RL         68.0     9717   Pave   NaN      Reg   
1459  1460          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  GdPrv        Shed    2500   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1459         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
0         2   2008        WD         Normal     208500  
1         5   2007        WD         Normal     181500  
2         9   2008        WD         Normal     223500  
3         2   2006        WD        Abnorml     140000  
4        12   2008        WD         Normal     250000  
...     ...    ...       ...            ...        ...  
1455      8   2007        WD         Normal     175000  
1456      2   2010        WD         Normal     210000  
1457      5   2010        WD         Normal     266500  
1458      4   2010        WD         Normal     142125  
1459      6   2008        WD         Normal     147500  

[1460 rows x 81 columns]

------------------------------------------------------------------------------------------


### Data Science Stage 3 :- Data Preprocessing ( i am taking the liberty of adding data cleaning in this stage ) 

In [85]:
for col in train_data.columns: # imputation method for train dataset
    if train_data[col].dtype in ['int64','float64']:
        train_data[col] = train_data[col].fillna(train_data[col].median())
    else:
        train_data[col] = train_data[col].fillna(train_data[col].mode()[0])

In [86]:
test_data = pd.read_csv("test.csv")

test_data

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

In [87]:
for col in test_data.columns:# imputation method for test dataset 
    if test_data[col].dtype in ['int64','float64']:
        test_data[col] = test_data[col].fillna(test_data[col].median())
    else:
        test_data[col] = test_data[col].fillna(test_data[col].mode()[0])
        

In [88]:
x = train_data.drop(columns='SalePrice')
y = train_data['SalePrice']

### Given below is the Encoding for categorical varaibles ( 43 )  for building a predictive model 

### Train Dataset Encoding


In [111]:
cat_col = x.select_dtypes(include=['object']).columns
num_col = x.select_dtypes(include=['int64','float64']).columns

enco = OneHotEncoder(drop='first',sparse_output=False ,handle_unknown ='ignore')

encoded_data = enco.fit_transform(x[cat_col])

encoded_columns = enco.get_feature_names_out(cat_col)

cat_encoded_data = pd.DataFrame(encoded_data,columns = encoded_columns,index = x.index)

x_encoded = x.drop(columns=cat_col)

x_encoded = pd.concat([x[num_col],cat_encoded_data], axis=1)

print(x_encoded.info())

print(x_encoded.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Columns: 245 entries, Id to SaleCondition_Partial
dtypes: float64(211), int64(34)
memory usage: 2.7 MB
None
   Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  SaleType_ConLw  \
0          2003       196.0         706  ...             0.0             0.0   
1          1976         0.0         978  ...             0.0             0.0   
2          2002       162.0         486  ...             0.0             0.0   
3          1970  

In [112]:
x_encoded # this is the encoded full training dataset 

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1455  1456          60         62.0     7917            6            5   
1456  1457          20         85.0    13175            6            6   
1457  1458          70         66.0     9042            7            9   
1458  1459          20         68.0     9717            5            6   
1459  1460          20         75.0     9937            5            6   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  \
0          2003          2003       196.0         706  ...             0.0   
1          1976          1976         0.0         978  ...             0.0   
2          2001          2002       162.0         486  ...             0.0   
3          1915          1970         0.0         216  ...             0.0   
4          2000          2000       350.0         655  ...             0.0   
...         ...           ...         ...         ...  ...             ...   
1455       1999          2000         0.0           0  ...             0.0   
1456       1978          1988       119.0         790  ...             0.0   
1457       1941          2006         0.0         275  ...             0.0   
1458       1950          1996         0.0          49  ...             0.0   
1459       1965          1965         0.0         830  ...             0.0   

      SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0                0.0           0.0           0.0          1.0   
1                0.0           0.0           0.0          1.0   
2                0.0           0.0           0.0          1.0   
3                0.0           0.0           0.0          1.0   
4                0.0           0.0           0.0          1.0   
...              ...           ...           ...          ...   
1455             0.0           0.0           0.0          1.0   
1456             0.0           0.0           0.0          1.0   
1457             0.0           0.0           0.0          1.0   
1458             0.0           0.0           0.0          1.0   
1459             0.0           0.0           0.0          1.0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                       0.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   
...                     ...                   ...                   ...   
1455                    0.0                   0.0                   0.0   
1456                    0.0                   0.0                   0.0   
1457                    0.0                   0.0                   0.0   
1458                    0.0                   0.0                   0.0   
1459                    0.0                   0.0                   0.0   

      SaleCondition_Normal  SaleCondition_Partial  
0                      1.0                    0.0  
1                      1.0                    0.0  
2                      1.0                    0.0  
3                      0.0                    0.0  
4                      1.0                    0.0  
...                    ...                    ...  
1455                   1.0                    0.0  
1456                   1.0                    0.0  
1457                   1.0      

### Test Dataset Encoding 

In [113]:
test_encoded_data = enco.transform(test_data[cat_col])

test_encoded_columns = enco.get_feature_names_out(cat_col)

y_encoded = pd.DataFrame(test_encoded_data,columns=test_encoded_columns, index=test_data.index)

y_encoded = pd.concat([test_data[num_col],y_encoded],axis=1)

In [114]:
y_encoded.info()
y_encoded.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Columns: 245 entries, Id to SaleCondition_Partial
dtypes: float64(219), int64(26)
memory usage: 2.7 MB


Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0  1461          20         80.0    11622            5            6   
1  1462          20         81.0    14267            6            6   
2  1463          60         74.0    13830            5            5   
3  1464          60         78.0     9978            6            6   
4  1465         120         43.0     5005            8            5   

   YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  \
0       1961          1961         0.0       468.0  ...             0.0   
1       1958          1958       108.0       923.0  ...             0.0   
2       1997          1998         0.0       791.0  ...             0.0   
3       1998          1998        20.0       602.0  ...             0.0   
4       1992          1992         0.0       263.0  ...             0.0   

   SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0             0.0           0.0           0.0          1.0   
1             0.0           0.0           0.0          1.0   
2             0.0           0.0           0.0          1.0   
3             0.0           0.0           0.0          1.0   
4             0.0           0.0           0.0          1.0   

   SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
0                    0.0                   0.0                   0.0   
1                    0.0                   0.0                   0.0   
2                    0.0                   0.0                   0.0   
3                    0.0                   0.0                   0.0   
4                    0.0                   0.0                   0.0   

   SaleCondition_Normal  SaleCondition_Partial  
0                   1.0                    0.0  
1                   1.0                    0.0  
2                   1.0                    0.0  
3                   1.0                    0.0  
4                   1.0                    0.0  

[5 rows x 245 columns]

## Note :- Remember that feature scaling is usefull for caluculating the distance based models everytime even if the data is equal tailed you still need to look at features such as salary / price which might be over shawdowing another feature such as age then we need to scale so that it doesnt produce biased model , but feature scaling is not required for tree based models because its not distance based models

-------------------------------------------------------

### Data Science Stage 4 :- EDA 

In [32]:
x.describe() # i chose the train data because it has the values that we need without the data encoding therefore it gives us the story of whats going on

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1460.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.104795   10516.828082     6.099315   
std     421.610009    42.300571    23.846996    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    70.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1460.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.492466   443.639726  ...   
std       1.112799    30.202904     20.645407   180.795612   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   165.250000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

## Observations

### 1 . The average condition of houses in iowa is 5 with the average quality of 6 and the average year at which the houses were built was 1971 and the average year of remodeling was in 1984 with average houses sold in the year of 2007 in the month of august with a average sale price of 180k$

### 2 . The minimum amount sales done in iowa was in the the 2006 , with lowest sale record going in the month of jan , the oldest house built in iowa was in 1872 & oldest remodelling done was in 1950 , the lowest sale is 34k$

### 3. The lowest overall condition is 1 , with lowest quality at 1 , while the highest condition is at 9 and highest overall condition is at 10 , the most recent house built was in year 2010 with the latest remodelling done in 2010 , while the highest selling house was worth money 755,000 $ 

### therefore these are primary observations i could draw , and there are factors that depeict the price of the sale now lets look towards the features 

In [97]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
import numpy as np

In [115]:
scaler = StandardScaler()

x_scaled = pd.DataFrame(scaler.fit_transform(x_encoded),columns = x_encoded.columns, index = x_encoded.index)
y_scaled = pd.DataFrame(scaler.fit_transform(y_encoded),columns = y_encoded.columns, index = y_encoded.index)

x_train,x_test,y_train,y_test = train_test_split(x_scaled,y,test_size=0.2,random_state =42)

y_train_log = np.log1p(y_train)
y_test_log = np.log1p(y_test)

model = LinearRegression()

model.fit(x_train,y_train_log)

LinearRegression()

In [116]:
y_pred_log= model.predict(x_test)
max_log_val = np.log(1e6)
min_log_val = np.log(1e4)
y_pred_log= np.clip(y_pred_log,min_log_val,max_log_val)
y_pred = np.expm1(y_pred_log)

In [117]:
y_pred

array([154266.97667421, 351880.48586007,  97969.85514526, 167513.73338763,
       318379.45754362,  79956.01164348, 245819.87413411, 148923.78153813,
        77033.63834881, 141061.55443267, 144536.23659706, 117907.46815094,
        79243.53128741, 216226.43627068, 178879.62145979, 134103.35293873,
       189060.4392984 , 136061.63345033, 107673.87620343, 214142.52847299,
       158586.83570028, 206950.23049527, 174434.42753497, 134759.49082886,
       200137.55038959, 149303.54083352, 190644.54008185, 107531.15539607,
       180190.35067831, 191190.63067561, 170122.00081259, 278336.26819536,
       209396.1944005 , 115702.43085374, 257135.68779896, 149660.00190968,
       147905.80797467, 195257.18384675, 325253.93366426, 107040.50169638,
       135555.28817967, 224770.66908462, 108425.83157911, 375031.682521  ,
       128186.58723291, 129269.04871706, 104865.09468672, 134417.41541553,
       472962.58590433, 137109.00844172, 118007.13435814, 213575.11220748,
       103897.4623604 , 3

In [118]:
from sklearn.metrics import mean_absolute_error,mean_squared_error, r2_score

In [119]:
mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
r = r2_score(y_test,y_pred)
rmsle = np.sqrt(mean_squared_error(y_test_log,y_pred_log))

print(f"Mean absolute error : {mae:.2f}")
print(f"Mean Squared Error : {mse:.2f}")
print(f"r2 : {r:.2f}")
print(rmsle)

Mean absolute error : 18294.31
Mean Squared Error : 2501984297.82
r2 : 0.67
0.18898173636363214


In [120]:
test_predictions = model.predict(y_scaled)

test_predictions = np.clip(test_predictions,min_log_val,max_log_val)

test_pred = np.expm1(test_predictions)

submission = pd.DataFrame({'Id':test_data['Id'] , 'SalePrice':test_pred})

submission.to_csv("submission.csv",index =False)
